pages/en/lb4/Application.md

```diff
@@ -0,0 +1,223 @@
+" %}
+
+```ts
+export class MyApplication extends Application {
+  constructor() {
+    super({
+      components: [
+        MagicSuite,
+      ],
+      servers: {
+        'public': RestServer,
+        'private': RestServer,
+      },
+      controllers: {
+        FooController,
+        BarController,
+        BazController,
+      }
+    });
+  }
+}
+```
+
+#### Components
+The components collection allows bulk binding of component constructors within
+your `Application` instance's context.
+
+For more information on how to make use of components,
+see [Using Components](http://loopback.io/doc/en/lb4/Using-components.html).
+
+#### Controllers
+Much like the components collection, the controllers collection allows bulk
+binding of [Controllers](http://loopback.io/doc/en/lb4/Controllers.html) to
+the `Application` context.
+
+#### Servers
+The servers collection is a Map of server constructors, whose keys are used
+as part of the server's binding name.
+
+```ts
+const app = new Application({
+  servers: {
+    'public': RestServer,
+    'private': RestServer,
+  },
+});
+```
+In the above example, the two server instances would be bound to the Application
+context under the keys `servers.public`, and `servers.private` respectively.
+
+
+
+## Tips for application setup
+Here are some tips to help avoid common pitfalls and mistakes.
+
+### Use unique bindings
+Use binding names that are prefixed with a unique string that does not overlap
+with loopback's bindings. As an example, if your application is built for
+your employer FooCorp, you can prefix your binding with `fooCorp`.
+```ts
+// This is unlikely to conflict with keys used by other component developers
+// or within loopback itself!
+app.bind('fooCorp.widgetServer.config').to(widgetServerConfig);
+```
+
+### Avoid use of `getSync`
+We provide the [`getSync`](http://apidocs.loopback.io/@loopback%2fcontext/#getSync)
+function for scenarios where you cannot asynchronously retrieve your bindings,
+such as in constructor bodies.
+
+However, the number of scenarios in which you must do this are limited, and you
+should avoid potential race conditions and retrieve your bindings asynchronously
+using the [`get`](http://apidocs.loopback.io/@loopback%2fcontext/#get) function
+whenever possible.
+
+### Do not create multiple Application instances
+{% include warning.html content="
+  LoopBack 4 was **not** designed to accommodate multiple Application instances
+  per module.
+  Correct behaviour is not guaranteed and **should not be expected**.
+" %}
+
+Many bindings defined by the framework itself are utilized under the
+assumption that there will only be one instance of an `Application` class.
+
+### Use caution with singleton binding scopes
+By default, bindings for controllers will instantiate a new instance whenever
+they are injected or retrieved from their binding. Your application should only
+set singleton binding scopes on controllers when it makes sense to do so.
+
```

Comment:
I don't think it's true for LoopBack 4. Most of internal services should NOT be assumed as singletons. When we bind a service class to a given context, new instances will be created. Even BindingScope.SINGLETON means it's a singleton within the context subtree.

Reply:
Umm, that sentence says "don't make your bindings singletons unless it makes sense to do it". I think we're in agreement?

Reply:
Weird. My comment was against `+### Do not create multiple Application instances` paragraph.

Reply:
If we're talking about that section, the issue is that decorators like `@api`, are creating metadata entries using `Injection.defineMetadata`, which is our singleton-scoped wrapper around `reflect-metadata`. If two instances of Application are both importing that same shared reference, will they not be polluting each other's "global" binding scope?

Reply:
Edited previous comment to correct some definitions and names, and to add a link to what I'm referring to.

Reply:
Decorators such as `@api` add metadata to a class/property/method/param. These metadata should be independent of the hosting app. Multiple app instances should be able to share the same metadata. App instance specific information need to be captured somewhere else outside the decorations. 

BTW, you should be able to create app1 and app2 to partition controllers/repositories. This scheme might be more useful for platforms such as appconnect that builds on top of LoopBack. For applications, one `app` is the case for most of the time, especially in the spirit of microservices.

Reply:
You can bind controllers to server-level context to achieve this.

My concern is that partitioning may occur in areas like controllers, but devs might be tempted to recycle common definitions of things like Models, which will definitely run afoul of this limitation since they also make use of the `Reflector` instance: https://github.com/strongloop/loopback-next/blob/8c7f474a72df23a1fa2d33bee9d36ba41031fd63/packages/repository/src/decorators/model.ts#L30

Reply:
I think the nuance of documenting this minefield is only going to open devs to problems down the road, and I don't see the value gained in doing so when compared to just separating them into different microservices.

Reply:
My point is that we don't have to take a position in this doc. IMO, it would be better to remove this paragraph.